### ISSUE TITLE
Automated sensor-quality flagging for BASEPROD traverses (Rust, zero dependencies, finds the events you exclude by hand)

### ISSUE THREAD

**Issue by philphauler** [spaceuma/baseprod#2]
Hey @levingerdes

Read your JIRS 2025 paper and the BASEPROD dataset paper in Scientific Data. Really well-done work, especially making the pre-exported corrected CSVs available directly in the archives.

Your dataset paper documents specific time windows to exclude from analysis (Table 3) and your JIRS paper notes that vibration and noise make it difficult to interpret the FT signals over entire traverses. I built an open-source tool that automates exactly that kind of quality check.

struktura (MIT/Apache-2.0, Rust) runs a structural-change scan on sensor recordings. It flags the moment a channel's statistical behavior shifts: calibration jumps, drift onset, stuck values, and sudden changes in the signal's scaling structure. Zero Python, zero training data, zero dependencies beyond libm. One command per traverse, processes all six FT legs plus IMU and FOG, finishes in under a second on a laptop.

https://github.com/koscak-labs/struktura

WHAT IT FOUND ON YOUR DATA

I ran it on every BASEPROD traverse I have extracted. Two things stood out:

1. Traverse 2023-07-20_20-01-38 (the 71 m evening traverse without cameras). At 858 s the tool flags a simultaneous structural change across all six FT legs:

shared events (>= 2 channels within +/- 1 block):
  block 83 (~859 s @99 Hz): FL_Fz(-0.21,z12) FR_Fz(-0.20,z9) CL_Fz(-0.19,z9)
                             CR_Fz(-0.25,z11) BL_Fz(-0.22,z10) BR_Fz(-0.22,z9)

All six legs, same 10-second block, z-scores 8.9 to 11.6. FOG yaw rate and IMU pitch/roll rate shift in the same window. GNSS speed, heading, altitude, linear accelerations, wheel commands and the slip ratio do not change. The kinematics rule out a speed or heading change. Position at 858 s is roughly UTM 623822 E, 4658168 N (30T).

This traverse is not in your labelled set. Do you know what was at ~40 m into that drive? A terrain boundary, a bogie event, a mode change?

2. Traverse 2023-07-20_18-12-05 (the labelled one). The tool picks up a shared event at 1438 s on FR and CL Force_Z, and a separate BL event at 1407 s. These would show up in your SVM pipeline as noise in the training windows if they fall inside a labelled segment.

WHAT IT SOLVES FOR YOUR PIPELINE

Your fts-assessment preprocessing exports 1-second statistical windows and feeds them to an SVM. That pipeline does not flag sensor quality issues before classification. Running struktura first on the raw corrected CSVs would:

- Flag the time windows you currently exclude by hand (Table 3 in your dataset paper). Instead of documenting which intervals to skip, the tool reports them automatically with z-scores and timestamps.
- Catch bogie encoder drift and PTU potentiometer drift. Your dataset paper documents both; the tool's drift detector flags the onset.
- Identify calibration shifts. You noted the original FTS calibration was lost and re-done with assumptions about uniform effects across channels. If the re-calibration left a step in any channel, the tool finds it.
- Clean the training data for your SVM. Sensor events inside labelled segments would confuse a terrain classifier. Flagging them first lets you exclude or re-label those windows before training.

It runs on the full recording at the original sample rate, not on 1-second windows, so it does not lose the 99.85% of the raw data that your windowed pipeline discards.

HOW TO RUN IT

cargo run --release --example baseprod_rover -- path/to/2023-07-20_20-01-38/

One command, one traverse directory. Output: per-channel sparkline of the block scaling exponent, any structural changes with z-scores and timestamps, and a shared-events line showing which channels changed together (distinguishing a whole-rover event from a single-sensor artifact).

WHAT I TRIED THAT DID NOT WORK

I also tested whether the DFA scaling exponent improves terrain classification on your labelled segments. On five development traverses it appeared to help (29/35 with sd + alpha vs 23/35 with sd alone), but on a reserved sixth traverse the gain reversed (6/9 vs 8/9 for sd alone). The feature responds to something in the recordings but does not transfer reliably across traverses for classification. I am not claiming a terrain classifier and I am not comparing to your 95.6% SVM.

WHAT I WOULD LIKE TO DO NEXT

- If you have traverses with known sensor events (a documented dropout, a calibration jump, a bogie failure), I would like to run the tool on them. The detection should catch those; if it does not, I want to fix it.
- The tool compiles to no_std embedded targets (ARM Cortex-M, allocation-free core). Your GNC architecture paper notes resource constraints on space-grade processors as a limiting factor. If a real-time sensor-quality check running alongside your navigation loop would be useful, I am happy to discuss integration.
- Your JIRS paper concludes that the case for FTS in wheeled mobile bases is "harder to argue" given cost/benefit. If automated quality flagging makes FTS data easier to interpret over full traverses, it might shift that trade-off.

One command, reproducible from the public repo, happy to help integrate it.

All the best
